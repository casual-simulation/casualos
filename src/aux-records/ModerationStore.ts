/**
 * Defines a class that is able to store moderation events.
 */
export interface ModerationStore {
    /**
     * Saves the given inst report.
     * @param report The report to save.
     */
    saveUserInstReport(report: UserInstReport): Promise<void>;
}

export interface UserInstReport {
    /**
     * The ID of the report.
     */
    id: string;

    /**
     * The name of the record that the inst exists in.
     * Null if the inst is public.
     */
    recordName: string | null;

    /**
     * The name of the inst.
     */
    inst: string;

    /**
     * The ID of the user that reported the inst.
     * Null if the user was not logged in when reporting the inst.
     */
    reportingUserId: string | null;

    /**
     * The IP address that the report came from.
     * Null if the IP address could not be determined (i.e. in the case that the report was automatically generated).
     */
    reportingIpAddress: string | null;

    /**
     * Whether the report was automatically generated.
     */
    automaticReport: boolean;

    /**
     * The user-provided explaination that the inst was reported for.
     */
    reportReasonText: string;

    /**
     * The user-provided reason category that the inst was reported for.
     */
    reportReason: ReportReason;

    /**
     * The URL that was reported.
     */
    reportedUrl: string;

    /**
     * The permalink that was reported.
     */
    reportedPermalink: string;

    /**
     * The unix timestamp in milliseconds when the report was created.
     */
    createdAtMs: number;

    /**
     * The unix timestamp in milliseconds when the report was last updated.
     */
    updatedAtMs: number;
}

export type ReportReason =
    | 'poor-performance'
    | 'spam'
    | 'harassment'
    | 'copyright-infringement'
    | 'obscene'
    | 'illegal'
    | 'other';
