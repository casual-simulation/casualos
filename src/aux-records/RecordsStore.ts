import { ComIdConfig, ComIdPlayerConfig } from './ComIdConfig';

/**
 * Defines an interface for objects that can store records.
 */
export interface RecordsStore {
    /**
     * Updates the given record.
     * @param record The record that should be updated.
     */
    updateRecord(record: Record): Promise<void>;

    /**
     * Adds the given record to the store.
     * @param record The record to add.
     */
    addRecord(record: Record): Promise<void>;

    /**
     * Gets the record with the given name.
     * @param name The name of the record.
     */
    getRecordByName(name: string): Promise<Record>;

    /**
     * Adds the given record key to the store.
     * @param key The key to add.
     */
    addRecordKey(key: RecordKey): Promise<void>;

    /**
     * Gets the record key for the given record name that has the given hash.
     * @param recordName The name of the record.
     * @param hash The scrypt hash of the key that should be retrieved.
     */
    getRecordKeyByRecordAndHash(
        recordName: string,
        hash: string
    ): Promise<RecordKey>;

    /**
     * Gets the list of records that the user with the given ID owns.
     *
     * If null or undefined, then this store does not support this method.
     *
     * @param ownerId The ID of the user that owns the records.
     */
    listRecordsByOwnerId(ownerId: string): Promise<ListedRecord[]>;

    /**
     * Gets the list of records that the studio with the given ID owns.
     *
     * If null or undefined, then this store does not support this method.
     *
     * @param studioId The ID of the studio that owns the records.
     */
    listRecordsByStudioId(studioId: string): Promise<ListedRecord[]>;

    /**
     * Gets the list of records that the studio with the given ID owns and that the user with the given ID has access to.
     * @param studioId The ID of the studio.
     * @param userId The ID of the user.
     */
    listRecordsByStudioIdAndUserId(
        studioId: string,
        userId: string
    ): Promise<ListedRecord[]>;

    /**
     * Adds the given studio to the store.
     * @param studio The studio to add.
     */
    addStudio(studio: Studio): Promise<void>;

    /**
     * Creates a new studio and adds the given user as an admin.
     * @param studio The studio to create.
     * @param adminId The ID of the admin user.
     */
    createStudioForUser(
        studio: Studio,
        adminId: string
    ): Promise<{
        studio: Studio;
        assignment: StudioAssignment;
    }>;

    /**
     * Updates the given studio.
     * @param studio The studio record that should be updated.
     */
    updateStudio(studio: Studio): Promise<void>;

    /**
     * Gets the studio with the given ID.
     * @param id The ID of the studio.
     */
    getStudioById(id: string): Promise<Studio>;

    /**
     * Gets the studio that has the given comId.
     * @param comId The comId of the studio.
     */
    getStudioByComId(comId: string): Promise<Studio>;

    /**
     * Gets the studio that has the given stripe customer ID.
     * @param customerId The stripe customer ID for the studio.
     */
    getStudioByStripeCustomerId(customerId: string): Promise<Studio>;

    /**
     * Gets the list of studios that the user with the given ID has access to.
     * Returns only studios that are not owned by any comId.
     * @param userId The ID of the user.
     */
    listStudiosForUser(userId: string): Promise<StoreListedStudio[]>;

    /**
     * Gets the list of studios that are owned by the given comId and that the user with the given ID has access to.
     * @param userId The ID of the user.
     * @param comId The comId of the studio that owns the studios.
     */
    listStudiosForUserAndComId(
        userId: string,
        comId: string
    ): Promise<StoreListedStudio[]>;

    /**
     * Gets the number of studios that are owned by the given comId.
     * @param comId The comId.
     */
    countStudiosInComId(comId: string): Promise<number>;

    /**
     * Adds the given studio assignment to the store.
     * @param assignment The assignment to add.
     */
    addStudioAssignment(assignment: StudioAssignment): Promise<void>;

    /**
     * Updates the given studio assignment.
     * @param assignment The assignment that should be updated.
     */
    updateStudioAssignment(assignment: StudioAssignment): Promise<void>;

    /**
     * Removes the given user from the given studio.
     * @param studioId The ID of the studio.
     * @param userId The ID of the user.
     */
    removeStudioAssignment(studioId: string, userId: string): Promise<void>;

    /**
     * Gets the list of users that have been assigned to the given studio.
     * @param studioId The ID of the studio.
     * @param filters The additional filters that should be used.
     */
    listStudioAssignments(
        studioId: string,
        filters?: ListStudioAssignmentFilters
    ): Promise<ListedStudioAssignment[]>;

    /**
     * Gets the list of studio assignments that the user with the given ID has access to.
     * @param userId The ID of the user.
     */
    listUserAssignments(userId: string): Promise<ListedUserAssignment[]>;

    /**
     * Counts the number of records that match the given filter.
     * @param filter The filter.
     */
    countRecords(filter: CountRecordsFilter): Promise<number>;

    /**
     * Saves the given comId request.
     * @param request The request.
     */
    saveComIdRequest(request: StudioComIdRequest): Promise<void>;
}

export interface CountRecordsFilter {
    /**
     * The ID of user that owns the record.
     */
    ownerId?: string;

    /**
     * The ID of the studio that owns the record.
     */
    studioId?: string;
}

/**
 * Defines an interface for record objects.
 */
export interface Record {
    /**
     * The name of the record.
     */
    name: string;

    /**
     * The ID of the user that owns the record.
     * Null if the record is owned by a studio.
     */
    ownerId: string | null;

    /**
     * The ID of the studio that owns the record.
     * Null if the record is owned by a user.
     */
    studioId: string | null;

    /**
     * The scrypt hashes of the secrets that allow access to the record.
     */
    secretHashes: string[];

    /**
     * The salt that is used to hash the secrets.
     *
     * Normally it is bad to share a salt between multiple secrets but in this case
     * it is fine because there are very few secrets per salt (i.e. not 1 salt per million users but 1 salt per couple record keys) and the secrets are randomly generated.
     */
    secretSalt: string;
}

export interface ListedRecord {
    /**
     * The name of the record.
     */
    name: string;

    /**
     * The ID of the user that owns the record.
     * Null if owned by a studio.
     */
    ownerId: string | null;

    /**
     * The ID of the studio that owns the record.
     * Null if owned by a user.
     */
    studioId: string | null;
}

/**
 * Defines an interface for studio objects.
 */
export interface Studio {
    /**
     * The ID of the studio.
     */
    id: string;

    /**
     * The name of the studio.
     */
    displayName: string;

    /**
     * The URL of the logo for the studio.
     */
    logoUrl?: string | null;

    /**
     * The ID of the stripe customer for this studio.
     */
    stripeCustomerId?: string;

    /**
     * The current subscription status for this studio.
     */
    subscriptionStatus?: string;

    /**
     * The ID of the purchasable subscription that the user has.
     * Note that this is the ID of the subscription in the config, not the ID of the stripe subscription.
     */
    subscriptionId?: string;

    /**
     * The ID of the subscription that this studio record references.
     */
    subscriptionInfoId?: string;

    /**
     * The unix time in miliseconds that the studio's current subscription period started at.
     */
    subscriptionPeriodStartMs?: number | null;

    /**
     * The unix time in miliseconds that the studio's current subscription period ends at.
     */
    subscriptionPeriodEndMs?: number | null;

    /**
     * The comId that this studio owns.
     */
    comId?: string | null;

    /**
     * The comId of the studio that owns this studio.
     */
    ownerStudioComId?: string | null;

    /**
     * The player web config for the studio.
     */
    playerConfig?: ComIdPlayerConfig;

    /**
     * The config for comId features.
     */
    comIdConfig?: ComIdConfig;
}

/**
 * Defines the list of possible studio roles that a user can be assigned.
 *
 * @dochash types/records/studios
 * @docname StudioAssignmentRole
 */
export type StudioAssignmentRole = 'admin' | 'member';

/**
 * Defines an interface for studio assignment objects.
 */
export interface StudioAssignment {
    /**
     * The ID of the studio that this assignment applies to.
     */
    studioId: string;

    /**
     * The ID of the user that this assignment applies to.
     */
    userId: string;

    /**
     * Whether the user is the primary contact for this studio.
     */
    isPrimaryContact: boolean;

    /**
     * The role that this user has in the studio.
     */
    role: StudioAssignmentRole;
}

export interface ListedStudioAssignment {
    /**
     * The ID of the studio that this assignment applies to.
     */
    studioId: string;

    /**
     * The ID of the user that this assignment applies to.
     */
    userId: string;

    /**
     * Whether the user is the primary contact for this studio.
     */
    isPrimaryContact: boolean;

    /**
     * The role that this user has in the studio.
     */
    role: StudioAssignmentRole;

    /**
     * The user that this assignment applies to.
     */
    user: ListedStudioAssignmentUser;
}

export interface ListedUserAssignment {
    /**
     * The name of the studio that this assignment applies to.
     */
    displayName: string;

    /**
     * The ID of the studio that this assignment applies to.
     */
    studioId: string;

    /**
     * The ID of the user that this assignment applies to.
     */
    userId: string;

    /**
     * Whether the user is the primary contact for this studio.
     */
    isPrimaryContact: boolean;

    /**
     * The role that this user has in the studio.
     */
    role: StudioAssignmentRole;
}

/**
 * The user information for a listed studio assignment.
 */
export interface ListedStudioAssignmentUser {
    /**
     * The ID of the user.
     */
    id: string;

    /**
     * The name of the user.
     */
    name: string;

    /**
     * The email address of the user.
     */
    email: string;

    /**
     * The phone number of the user.
     */
    phoneNumber: string;
}

/**
 * Defines an interface that represents a studio that a user has access to.
 */
export interface StoreListedStudio {
    /**
     * The ID of the studio.
     */
    studioId: string;

    /**
     * The name of the studio.
     */
    displayName: string;

    /**
     * The URL of the logo for the studio.
     */
    logoUrl: string;

    /**
     * The role that the user has in the studio.
     */
    role: StudioAssignmentRole;

    /**
     * Whether the user is the primary contact for this studio.
     */
    isPrimaryContact: boolean;

    /**
     * The ID of the studio's subscription.
     */
    subscriptionId: string;

    /**
     * The current subscription status for this studio.
     */
    subscriptionStatus: string;

    /**
     * The comId of the studio.
     */
    comId: string | null;

    /**
     * The comId of the studio that owns this studio.
     */
    ownerStudioComId: string | null;
}

export interface ListStudioAssignmentFilters {
    /**
     * The ID of the user to filter by.
     */
    userId?: string;

    /**
     * The role to filter by.
     */
    role?: string;

    /**
     * Whether to filter by primary contact.
     */
    isPrimaryContact?: boolean;
}

/**
 * Defines a type that represents the different kinds of policies that a record key can have.
 *
 * - null and "subjectfull" indicate that actions performed with this key must require a subject to provide their access token in order for operations to succeed.
 * - "subjectless" indicates that actions may be performed with key despite not having an access key from a subject.
 *
 * @dochash types/records/key
 * @docname PublicRecordKeyPolicy
 */
export type PublicRecordKeyPolicy = null | 'subjectfull' | 'subjectless';

/**
 * Defines an interface for record key objects.
 */
export interface RecordKey {
    /**
     * The name of the record that the key is for.
     */
    recordName: string;

    /**
     * The scrypt hash of the secret that this key is for.
     */
    secretHash: string;

    /**
     * The policy that the key uses.
     */
    policy: PublicRecordKeyPolicy;

    /**
     * The ID of the user that created this key.
     */
    creatorId: string;
}

/**
 * Defines an interface that represents a request for a comId to be applied to a studio.
 */
export interface StudioComIdRequest {
    /**
     * The ID of the request.
     */
    id: string;

    /**
     * The ID of the studio.
     */
    studioId: string;

    /**
     * The ID of the user that is making the request.
     * Null if the user has been deleted.
     */
    userId: string;

    /**
     * The comId that is being requested.
     */
    requestedComId: string;

    /**
     * The IP Address that the request came from.
     */
    requestingIpAddress: string;

    /**
     * The unix timestamp in miliseconds when the request was created.
     */
    createdAtMs: number;

    /**
     * The unix timestamp in miliseconds when the request was last updated.
     */
    updatedAtMs: number;
}
