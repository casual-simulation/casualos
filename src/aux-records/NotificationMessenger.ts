import { UserInstReport } from './ModerationStore';
import { DateTime } from 'luxon';
import { z } from 'zod';

/**
 * Defines an interface for a class that is able to send records notifications.
 */
export interface NotificationMessenger {
    /**
     * Sends a notification to the user.
     * @param notification The notification to send.
     */
    sendRecordNotification(notification: RecordsNotification): Promise<void>;
}

/**
 * Defines a class that is able to send notifications to multiple messengers.
 */
export class MultiNotificationMessenger implements NotificationMessenger {
    private _messengers: NotificationMessenger[];
    private _options: NotificationOptions;

    constructor(options: NotificationOptions) {
        this._options = options;
        this._messengers = [];
    }

    addMessenger(messenger: NotificationMessenger) {
        this._messengers.push(messenger);
    }

    async sendRecordNotification(
        notification: UserInstReportNotification
    ): Promise<void> {
        if (this._options.filter) {
            const filter = this._options.filter;
            if (
                filter.resources &&
                !filter.resources.includes(notification.resource)
            ) {
                return;
            }

            if (
                filter.actions &&
                !filter.actions.includes(notification.action)
            ) {
                return;
            }
        }

        await Promise.all(
            this._messengers.map((m) => m.sendRecordNotification(notification))
        );
    }
}

export type RecordsNotification = UserInstReportNotification;

/**
 * Defines a base interface for a notification that is related to a record.
 */
export interface RecordsNotificationBase {
    /**
     * The name of the record that the notification is for.
     * Null or undefined if the notification is not for a specific record.
     */
    recordName?: string;

    /**
     * The unix time in milliseconds that the notification was sent.
     */
    timeMs: number;
}

export interface ResourceNotification extends RecordsNotificationBase {
    /**
     * The action that was taken on the resource.
     */
    action: NotificationResourceActions;

    /**
     * The ID of the resource.
     */
    resourceId: string;
}

/**
 * Defines a notification that is for a user inst report.
 */
export interface UserInstReportNotification extends ResourceNotification {
    resource: 'user_inst_report';

    /**
     * The report that was created.
     */
    report: UserInstReport;
}

export type NotificationResourceActions = 'created' | 'updated' | 'deleted';

export const slackSchema = z.object({
    webhookUrl: z
        .string()
        .describe(
            'The Slack webhook URL that should be used to send records notification messages.'
        )
        .url(),
});

export type SlackOptions = z.infer<typeof slackSchema>;

export const telegramSchema = z.object({
    chatId: z
        .number()
        .describe(
            'The ID of the Telegram chat that messages should be sent to.'
        ),
    token: z
        .string()
        .describe(
            'The Telegram bot token that should be used to send messages.'
        ),
});

export type TelegramOptions = z.infer<typeof telegramSchema>;

export const notificationFilterSchema = z.object({
    resources: z
        .array(z.string())
        .describe(
            'The resources that match the filter. If omitted, then all resources are matched.'
        )
        .optional(),
    actions: z
        .array(z.string())
        .describe(
            'The actions that match the filter. If omitted, then all actions are matched.'
        )
        .optional(),
});

export const notificationsSchema = z.object({
    slack: slackSchema
        .describe(
            'The Slack configuration that should be used for notifications. If omitted, then notifications will not be sent via Slack.'
        )
        .optional(),
    telegram: telegramSchema
        .describe(
            'The Telegram configuration that should be used for notifications. If omitted, then notifications will not be sent via Telegram.'
        )
        .optional(),
    filter: notificationFilterSchema
        .describe(
            'The filter that should be used to determine which notifications should be sent. If omitted, then all are sent'
        )
        .optional()
        .default({}),
});

export type NotificationOptions = z.infer<typeof notificationsSchema>;

export function formatNotificationAsString(
    notification: RecordsNotification
): string {
    switch (notification.resource) {
        case 'user_inst_report':
            return formatUserInstReportNotificationAsString(notification);
    }
}

export function formatUserInstReportNotificationAsString(
    notification: UserInstReportNotification
): string {
    const time = DateTime.fromMillis(notification.timeMs, {
        zone: 'utc',
    }).toISO();
    return `A user inst report was ${notification.action} for ${
        notification.recordName ?? '(null)'
    }/${notification.resourceId}:
Report ID: ${notification.report.id}
Time: ${time}
Reporting User: ${notification.report.reportingUserId ?? '(null)'}
Reporting IP: ${notification.report.reportingIpAddress ?? '(null)'}
Reported URL: ${notification.report.reportedUrl}
Reported Permalink: ${notification.report.reportedPermalink}
Is Automatic Report?: ${notification.report.automaticReport}

Reason: ${notification.report.reportReason}

Text:
${notification.report.reportReasonText}`;
}
