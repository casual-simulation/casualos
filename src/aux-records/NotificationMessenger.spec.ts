import { UserInstReport } from './ModerationStore';
import { formatNotificationAsString } from './NotificationMessenger';
import { StudioComIdRequest } from './RecordsStore';

describe('formatNotificationAsString()', () => {
    it('should consistently format user inst report notifications', () => {
        const report: UserInstReport = {
            id: 'test_id',
            createdAtMs: 123,
            updatedAtMs: 123,
            recordName: 'test_record',
            inst: 'test_inst',
            reportedPermalink: 'test_permalink',
            reportedUrl: 'test_url',
            reportReason: 'harassment',
            reportingIpAddress: '127.0.0.1',
            automaticReport: false,
            reportingUserId: 'userId',
            reportReasonText: 'test_reason',
        };

        const result = formatNotificationAsString({
            resource: 'user_inst_report',
            action: 'created',
            resourceId: 'test_id',
            timeMs: 123,
            report: report,
        });
        expect(result).toMatchSnapshot();
    });

    it('should consistently format studio comId request notifications', () => {
        const request: StudioComIdRequest = {
            id: 'test_id',
            createdAtMs: 123,
            updatedAtMs: 123,
            requestedComId: 'test_comId',
            requestingIpAddress: '127.0.0.1',
            studioId: 'test_studioId',
            userId: 'test_userId',
        };

        const result = formatNotificationAsString({
            resource: 'studio_com_id_request',
            action: 'created',
            resourceId: 'test_id',
            timeMs: 123,
            request: request,
        });
        expect(result).toMatchSnapshot();
    });
});
