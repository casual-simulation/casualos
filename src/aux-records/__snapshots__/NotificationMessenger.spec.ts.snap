// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatNotificationAsString() should consistently format studio comId request notifications 1`] = `
"A comId request was created for test_id:
Request ID: test_id
Studio ID: test_studioId
Requested comID: test_comId
Time: 1970-01-01T00:00:00.123Z
Reporting User: test_userId
Requesting IP: 127.0.0.1"
`;

exports[`formatNotificationAsString() should consistently format user inst report notifications 1`] = `
"A user inst report was created for (null)/test_id:
Report ID: test_id
Time: 1970-01-01T00:00:00.123Z
Reporting User: userId
Reporting IP: 127.0.0.1
Reported URL: test_url
Reported Permalink: test_permalink
Is Automatic Report?: false

Reason: harassment

Text:
test_reason"
`;
